user_app/bin/menu/base/field.dart

```diff
@@ -0,0 +1,42 @@
+import '../../exceptions/course_already_added_exception.dart';
+import '../../exceptions/input_exception.dart';
+import '../../exceptions/invalid_choice_exception.dart';
+import '../../exceptions/roll_number_already_exists_exception.dart';
+import '../../utils/logger.dart';
+
+abstract class Field<T> {
+  final String name;
+  // can be protected but no option in dart
+  final String message;
+
+  late T? value;
+
+  Field(this.name, this.message);
+
+  T? readValue();
+
+  /// catches all the [InputException] and logs it according to the log level.
+  bool catchError(Function block) {
+    try {
+      block();
+      return true;
+    } on RollNumberAlreadyExistsException catch (e) {
+      Log.error(e);
+      Log.help(
+          "The roll number that you have entered is already present in the file storage.");
+    } on InvalidChoiceException catch (e) {
+      Log.error(e);
+      Log.help("Enter the option as specified by the prompt.");
+    } on CourseAlreadyAddedException catch (e) {
+      Log.error(e);
+      Log.help("You have already entered the same course before.");
+    } on InputException catch (e, stackTrace) {
+      Log.error(e);
+      Log.error(stackTrace);
+    } catch (e, stackTrace) {
+      Log.debug(e);
+      Log.debug(stackTrace);
+    }
```

Comment:
If you are throwing custom exceptions from the block() then I guess the message is a constant for each type right ? 
So you can give a default message for each exception in the implementation of the exception and accept a message as optional input. 
So you can just end up overriding the message only wherever required. 
And for the rest of the exceptions you can catch the base exception and log the toString() of it 

For example : 

```
try{
    throw IllegalTypeException();
  } on CollegeAppException catch(exception,stackTrace){
    print(exception.toString());
    print(stackTrace);
  }
```

So CollegeAppException is the base exception here and IllegalTypeException is a custom exception that extends CollegeAppException.